Allow empty condition for jsonpath

aparekh7 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
Extends current JSONPath condition logic to support matching "any" value for the provided JSON path expression.
With this change, users will be able to wait on JSON path conditions with values they cannot know ahead including simple or complex (object or array) values. See https://github.com/kubernetes/kubernetes/issues/117761 for more details.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #117761 

#### Special notes for your reviewer:
This is my first contribution! Please go easy on me haha

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
The "value" part in the `wait --for=jsonpath='{expression}'[=value]` is now optional. If the value is not provided i.e. the command looks like `wait --for=jsonpath='{expression}'` then the wait condition is interpreted as matched when the expression returns *any* single JSON value like object or a literal.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

parekhaayush committed:
Allow empty condition for jsonpath

aparekh7 commented:
/check-cla

aparekh7 commented:
/check-cla

aparekh7 commented:
/easycla

minherz commented on staging/src/k8s.io/kubectl/pkg/cmd/wait/wait.go:
```diff
@@ -207,10 +207,14 @@ func conditionFuncFor(condition string, errOut io.Writer) (ConditionFunc, error)
+			return nil, fmt.Errorf("jsonpath wait format must be --for=jsonpath='{.status.readyReplicas}'[=3]")
 		}
-		jsonPathExp, jsonPathCond, err := processJSONPathInput(splitStr[1], splitStr[2])
+		conditionStr := ""
+		if len(splitStr) == 3 {
+			conditionStr = splitStr[2]
+		}
+		jsonPathExp, jsonPathCond, err := processJSONPathInput(splitStr[1], conditionStr)
```
nit: refactoring it by explicitly handling a case without "JSONPath condition" will reduce cognitive load when reading the code. I suggest to handle the cases `len(splitStr) == 2` and `len(splitStr) == 3` both for readability and code cleaness

aparekh7 replied:
Done

minherz commented on staging/src/k8s.io/kubectl/pkg/cmd/wait/wait.go:
```diff
@@ -667,5 +668,5 @@ func compareResults(r reflect.Value, expectedVal string) (bool, error) {
 		return false, errors.New("jsonpath leads to a nested object or list which is not supported")
 	}
 	s := fmt.Sprintf("%v", r.Interface())
-	return strings.TrimSpace(s) == strings.TrimSpace(expectedVal), nil
+	return strings.TrimSpace(s) == strings.TrimSpace(expectedVal) ||  expectedVal == "", nil
```
Does the code guarantee that `s` is not empty? When the value is not provided in the wait for jsonpath, the expectation is that wait successfully completes on _any non empty_ value.

aparekh7 replied:
From what I have understood from the code, it does seem like `s` cannot be empty. However, it would be great if someone with experience could confirm this!
Have added a check for this though

minherz replied:
According to the [signature] it can be empty string in an event `r.Interface()` returns an empty string.

[signature]: https://pkg.go.dev/reflect#Value.Interface

minherz commented on staging/src/k8s.io/kubectl/pkg/cmd/wait/wait.go:
```diff
@@ -667,5 +668,5 @@ func compareResults(r reflect.Value, expectedVal string) (bool, error) {
 		return false, errors.New("jsonpath leads to a nested object or list which is not supported")
```
According to the requirement, when JSONPath condition is not defined, this should return `true` because the object fits to the definition of the "_any non empty_" value.

aparekh7 replied:
A nested object or list is technically a "_non-empty value_", but it is still something whose evaluation is not supported currently. I feel it would be incorrect here to return `true` in this scenario. If we are to look at it this way, won't that render the verifications done in `verifyParsedJSONPath()`  invalid? This function raises errors when the  `jsonpath` expression matches more than one list/value, which is also technically a "_non-empty value_"
Maybe this can be done once nested objects/lists and multiple values are supported to wait on

minherz replied:
The feature request's requirement is to allow omitting a JSON value and when the value is not provided, to wait for a JSON expression to have a non-empty value. That value should not be further evaluated or processed. Determination of the value as "non-empty" is a stopping condition. Using the original example with a service of the type `LoadBalancer`, there is no need to wait for `.status.loadBalancer.ingress[:].ip`. It will be sufficient to wait for `.status.loadBalancer`.
The unsupported part is for the use case when a caller _provides_ a value. In this case, only literals are supported.

aparekh7 replied:
Got it.
Is it safe to assume that if the `jsonpath` expression evaluates to any value which is not `""`, it is a "non-empty value"?

minherz replied:
I believe it is. This is how the value was interpreted so far and is defined in JSON standard AFAIK

minherz commented on staging/src/k8s.io/kubectl/pkg/cmd/wait/wait.go:
```diff
@@ -246,9 +250,6 @@ func processJSONPathInput(jsonPathExpression, jsonPathCond string) (string, stri
 	if err != nil {
 		return "", "", err
 	}
-	if jsonPathCond == "" {
-		return "", "", errors.New("jsonpath wait condition cannot be empty")
-	}
```
Removing this code introduces a bug when the command is called in the form of `--for=jsonpath='{.status.readyReplicas}'=` i.e. without the value after second `=` sign.

aparekh7 replied:
Handled this case

minherz reviewed (changes requested):
Documentation should be modified to emphasis that "JSONPath Condition" is optional.
I think that "JSONPath Condition" worth renaming because "condition" may hint that the variable is an expression when it is just a literal value.

minherz commented:
Please, edit documentation in 
https://github.com/kubernetes/kubernetes/blob/5fd29596ef699a8658c941c1c7cf7a45e796c1c6/staging/src/k8s.io/kubectl/pkg/cmd/wait/wait.go#L123

minherz commented:
@aparekh7 apologies for bothering, can you share your plans on addressing this PR?

aparekh7 commented:
@minherz i got busy with my job, i plan to address these comments in the next few days

parekhaayush committed:
Allow empty condition for jsonpath

- Rename jsonPathCondition to jsonPathExpectedResult
- Edit documentation to denote optional expected result for jsonpath
- Fix bugs introduced
- Refactored code

aparekh7 commented:
/check-cla
/easycla

minherz commented on staging/src/k8s.io/kubectl/pkg/cmd/wait/wait_test.go:
```diff
@@ -1046,6 +1046,18 @@ func TestWaitForDifferentJSONPathExpression(t *testing.T) {
 
 			expectedErr: "timed out waiting for the condition on theresource/foo-b6699dcfb-rnv7t",
 		},
+		{
+			name: "JSONPath condition not given",
+			fakeClient: func() *dynamicfakeclient.FakeDynamicClient {
+				fakeClient := dynamicfakeclient.NewSimpleDynamicClientWithCustomListKinds(scheme, listMapping)
+				fakeClient.PrependReactor("list", "theresource", listReactionfunc)
+				return fakeClient
+			},
+			jsonPathExp:  "{.status.containerStatuses[0].ready}",
+			jsonPathCond: "",
+
+			expectedErr: None,
+		},
```
nit: Consider adding a test for "empty" value that is currently not implemented (i.e. `{.status.containerStatuses[0].ready}=`).

minherz replied:
Yes, there is a clearly missing test that validates the correct parsing. It should handle the expected error when passing "invalid" json expression

minherz reviewed (changes requested):
Thank you for addressing my remarks. It seems that the recent changes introduced a couple of issues. Please have a look into my comments.
I'd like to remark on your decision of splitting `processJSONPathInput` function into two separate methods. For me, this change added cognitive load to understand the logic. The reasons for this is passing almost same input to both functions that called sequentially and that the logic in both mutually related.
It is possible that the code would be easier to understand if the input of the JSON condition will be processed in one place IMHO.

minherz commented:
@aparekh7, it looks like you did not sign Kubernetes CLA. Please, note that you cannot contribute to the project without signing it. You can look into details of the EasyCLA failed check for more info.

mrbobbytables commented:
/easycla

minherz commented:
@aparekh7, I'd like to follow up with you about a possibility to finalize this PR before the code freeze that is planned to happened within next 3 weeks. Based on the current posture there is some work remained. If your capacity to work on this PR is limited, I would like to take over to implement this issue. Please, let me know if you would like to discuss this over GVC/Skype/Zoom or on Slack/Chat.

aparekh7 commented on staging/src/k8s.io/kubectl/pkg/cmd/wait/wait.go:
```diff
@@ -210,11 +210,11 @@ func conditionFuncFor(condition string, errOut io.Writer) (ConditionFunc, error)
+			return nil, err
 		}
-		jsonPathExp, jsonPathCond, err := processJSONPathInput(splitStr[1], conditionStr)
+		jsonPathExp, err := processJSONPathExpression(splitStr[1])
```
Yes, the signature of `processJSONPathExpression` is `func processJSONPathExpression(jsonPathExpression string) (string, error)` and not `func processJSONPathExpression([]string) (string, error)`

parekhaayush committed:
Allow empty condition for jsonpath

- Reformat code
- Fix bugs

aparekh7 commented:
@minherz I have signed the CLA, I also received an email saying the same. I am not sure why that status is not reflected here. If there is some issue with this PR, I will close it and open a new one once these changes are fine

minherz commented on staging/src/k8s.io/kubectl/pkg/cmd/wait/wait.go:
```diff
@@ -240,18 +237,55 @@ func newJSONPathParser(jsonPathExpression string) (*jsonpath.JSONPath, error) {
-// processJSONPathInput will parses the user's JSONPath input and process the string
-func processJSONPathInput(jsonPathExpression, jsonPathCond string) (string, string, error) {
-	relaxedJSONPathExp, err := cmdget.RelaxedJSONPathExpression(jsonPathExpression)
+func processJSONPathInput(jsonPathSplitStr []string) (string, string, error) {
```
nit: It is hard to understand the logic behind this code. Consider to simplify by compartmentalizing tasks and keeping naming of functions clear. For example, `validateJSONPathInput` just checks the number of arguments. Unless it validates that 2nd argument is a valid JSON and 3rd, if provided, is not empty, the "validate" does not cover all validations. You can either rename it to narrow the scope of what the function do or to modify the method to complete all "validate" tasks.
It is a good practice to avoid functions that just forward the call results of the nested function like `processJSONPathExpression`

minherz commented on staging/src/k8s.io/kubectl/pkg/cmd/wait/wait.go:
```diff
@@ -632,16 +666,27 @@ func (j JSONPathWait) checkCondition(obj *unstructured.Unstructured) (bool, erro
+		return true
+	}
+	result := fmt.Sprintf("%v", results[0][0].String())
+	return result != ""
```
Consider replacing it with:
```go
return strings.TrimSpace(result) != ""
```
like it is done in `compareResults()`

minherz commented:
/test

lauchokyip commented:
/ok-to-test

parekhaayush committed:
Allow empty condition for jsonpath
- Reformat code
- Add tests to validate JSONPath input

minherz commented:
/easycla

minherz commented:
The changes were merged in #118160 
/close
